pkg/virt-controller/watch/vm.go

```diff
@@ -1126,6 +1129,26 @@ func (c *VMController) setupVMIFromVM(vm *virtv1.VirtualMachine) *virtv1.Virtual
 	return vmi
 }
 
+// Ensure each volume has an associated disk or filesystem and if not create a default disk.
+func setupMissingVolumeDisks(vmi *virtv1.VirtualMachineInstance) {
+
+	diskAndFilesystemNames := util.FetchDiskAndFSNames(vmi)
+
+	for _, volume := range vmi.Spec.Volumes {
+		if _, foundDisk := diskAndFilesystemNames[volume.Name]; !foundDisk {
+			vmi.Spec.Domain.Devices.Disks = append(
+				vmi.Spec.Domain.Devices.Disks,
+				virtv1.Disk{
+					Name: volume.Name,
+					DiskDevice: virtv1.DiskDevice{
+						Disk: &virtv1.DiskTarget{},
+					},
```

Comment:
what is the default bus?

Reply:
> One thing that catches my eye here is that VM and VMIs will get treated differently.
> 
> A VM without a disk matching a volume in the VMITemplate will have one created for it when the VM controller creates the VMI
> 
> However, a VMI posted standalone without a matching disk will fail.

Correct and honestly I was reading between the lines a little here with regards to the differences between the two. My assumption was that given the current behaviour of the flavors API that takes a reference from the VM and applies it in full to the VMI that we could treat them as separate APIs almost. The VM having various abstractions while the VMI has a much more strict API.

> What would you think about making the disk defaulting something that is implied within the VMI... meaning a volume without a matching disk has an "implied" disk with some default settings applied to it. In that scenario someone would only need to create the disk in order to override the defaults.

We could do this but I'd rather keep the checks in the VMI admitter in place and just handle this in the controller tbh.

> Also, i see how this definitely simplifies our api a bit, but i'm unsure how this will technically work with the `bus` fields. how would we correctly pick the bus? that's part of what flavors were goign to help with by giving us a `preferedDiskBus` hint.

With my current preference implementation in https://github.com/kubevirt/kubevirt/pull/7578 we would apply any disk bus preference *after* the logic I've added here but still before submitting the VMI request. If we don't have a disk bus preference then the original defaults are applied during submission, these currently are a DiskTarget [1] and SATA bus [2].

[1] https://github.com/kubevirt/kubevirt/blob/52e6e9bdd42c3cd727ba989c3dad0dd169f8259d/staging/src/kubevirt.io/api/core/v1/defaults.go#L69-L75
[2]  https://github.com/kubevirt/kubevirt/blob/52e6e9bdd42c3cd727ba989c3dad0dd169f8259d/staging/src/kubevirt.io/api/core/v1/defaults.go#L127-L148

Reply:
> One thing that catches my eye here is that VM and VMIs will get treated differently.
> 
> A VM without a disk matching a volume in the VMITemplate will have one created for it when the VM controller creates the VMI
> 
> However, a VMI posted standalone without a matching disk will fail.

Correct and honestly I was reading between the lines a little here with regards to the differences between the two. My assumption was that given the current behaviour of the flavors API that takes a reference from the VM and applies it in full to the VMI that we could treat them as separate APIs almost. The VM having various abstractions while the VMI has a much more strict API.

> What would you think about making the disk defaulting something that is implied within the VMI... meaning a volume without a matching disk has an "implied" disk with some default settings applied to it. In that scenario someone would only need to create the disk in order to override the defaults.

We could do this but I'd rather keep the checks in the VMI admitter in place and just handle this in the controller tbh.

> Also, i see how this definitely simplifies our api a bit, but i'm unsure how this will technically work with the `bus` fields. how would we correctly pick the bus? that's part of what flavors were goign to help with by giving us a `preferedDiskBus` hint.

With my current preference implementation in https://github.com/kubevirt/kubevirt/pull/7578 we would apply any disk bus preference *after* the logic I've added here but still before submitting the VMI request. If we don't have a disk bus preference then the original defaults are applied during submission, these currently are a DiskTarget [1] and SATA bus [2].

[1] https://github.com/kubevirt/kubevirt/blob/52e6e9bdd42c3cd727ba989c3dad0dd169f8259d/staging/src/kubevirt.io/api/core/v1/defaults.go#L69-L75
[2]  https://github.com/kubevirt/kubevirt/blob/52e6e9bdd42c3cd727ba989c3dad0dd169f8259d/staging/src/kubevirt.io/api/core/v1/defaults.go#L127-L148

Reply:
> What would you think about making the disk defaulting something that is implied within the VMI... meaning a volume without a matching disk has an "implied" disk with some default settings applied to it. In that scenario someone would only need to create the disk in order to override the defaults.

Sounds good to me.
Do you think it's reasonable to use a label to imply the right defaults? or maybe some VMI dedicated field? adding a field for every default disk configuration that's being overriden is much easier and less verbose than adding some disks to the manifest

Reply:
> > What would you think about making the disk defaulting something that is implied within the VMI... meaning a volume without a matching disk has an "implied" disk with some default settings applied to it. In that scenario someone would only need to create the disk in order to override the defaults.

Coming back to this, you would need to define the disk as a diskDevice *and* provide a bus to override the current defaults (as well as any preferences introduced in the future). I'd still rather keep this logic in the VM controller and leave the current strict API behaviour in place for the VMI.

> Sounds good to me. Do you think it's reasonable to use a label to imply the right defaults? or maybe some VMI dedicated field? adding a field for every default disk configuration that's being overriden is much easier and less verbose than adding some disks to the manifest

Flavor Preferences (https://github.com/kubevirt/kubevirt/pull/7578) are looking to solve this but at the moment as with this logic they are only applied by the VM controller when we launch a VM and initially create the VMI.

Reply:
> I'll look deeper into Flavor Preferences PR for sure.

Awesome thanks!

> Another thought in the meantime: the flavor preferences can be simply reflected by the VMI, then the VMI controller can handle the defaults and preferences instead of the VM controller. For example, the VM controller can add certain metadata labels about these defaults and preferences to the VMI template. Then, the VMI controller can be the one who actually performs the logic of configuring the disks (and other stuff) according to these labels. This way you gain more automation by using a VM, but you can still use a VMI and have the same capabilities. (of course we don't have to use labels, it could be something else).

At present we have matchers for both the flavor and preferences that the controller uses to lookup and apply things to the VMI before we submit it.

> WDYT? Does it make sense to you?

I've got a few concerns but I'm likely missing context. Do we expect users to create VMIs directly? Additionally is it acceptable for us to heavily modify resources after submission like this?

Reply:
> At present we have matchers for both the flavor and preferences that the controller use to lookup and apply things to the VMI before we submit it.

Applying to the VMI could be through labels or any other way that will hint the VMI controller on how to actually apply the settings to the target pod. This way VM controller only serves as automation that would grant metadata for the VMI controller. That would grant the possibility to use the same features on VMs and VMIs.

> I've got a few concerns but I'm likely missing context. Do we expect users to create VMIs directly?

Yes, creating VMIs is expected and it's a valid way of using Kubevirt. This is one of David's main concerns, IIUC. As he wrote:
```
 However, a VMI posted standalone without a matching disk will fail.
```

> Additionally is it acceptable for us to heavily modify resources after submission like this?

Heavily modify which resource, the VMI? To which heavy modification are you referring - the injection of new disks to the VMI spec? How is it that different from this PR changes?

If a VMI is created by creating a VM -> there is no modification since we're the ones creating the VMI
if the VMI is created directly -> we will only set default disks according to the defaults the VM creator set for us

Reply:
First of all, thanks for the continued comments @iholder-redhat ! 

> > At present we have matchers for both the flavor and preferences that the controller use to lookup and apply things to the VMI before we submit it.
> 
> Applying to the VMI could be through labels or any other way that will hint the VMI controller on how to actually apply the settings to the target pod. This way VM controller only serves as automation that would grant metadata for the VMI controller. That would grant the possibility to use the same features on VMs and VMIs.

Labels might work for a few items but I'm looking at introducing a large number of preferences so this might not scale well with what I have in mind :

https://github.com/kubevirt/kubevirt/blob/bde04005a0d5debb1a166934cf57cd62544a73c3/staging/src/kubevirt.io/api/flavor/v1alpha1/types.go#L183-L409

> > I've got a few concerns but I'm likely missing context. Do we expect users to create VMIs directly?
> 
> Yes, creating VMIs is expected and it's a valid way of using Kubevirt. This is one of David's main concerns, IIUC. As he wrote:
> 
> ```
>  However, a VMI posted standalone without a matching disk will fail.
> ```

ACK thanks, I'm still not convinced by this concern if I'm honest. Yes you would get different behaviour but these are after all different CRs. Wrapping VirtualMachine with some automation like this (and flavors, preferences etc) while leaving the runtime VirtualMachineInstance CR be more strict still seems like the right thing to do for me.

> > Additionally is it acceptable for us to heavily modify resources after submission like this?
> 
> Heavily modify which resource, the VMI? To which heavy modification are you referring - the injection of new disks to the VMI spec? How is it that different from this PR changes?

Yes the additional disks in this case but I'm also thinking in general of the application of preferences to the VMI post submission that I think you were also suggesting. The disks are added to the VMI prior to submission in this PR so the VMI resource itself doesn't change after submission.

> If a VMI is created by creating a VM -> there is no modification since we're the ones creating the VMI if the VMI is created directly -> we will only set default disks according to the defaults the VM creator set for us

In both cases the VMI resource is still going to be mutated after submission right? Even if we do that at the start of the VMI admittance process the definition supplied by the user or VM controller still isn't going to reflect what actually ends up running. Yes it will have references to things we apply but it isn't the same. 

The reason I'm standing by the current proposal is because the users VM definition and the VMI definition created by the VM controller ultimately don't change after being admitted. We also leave all of the VMI validations in place to ensure that it can actually launch correctly.

Reply:
> First of all, thanks for the continued comments @iholder-redhat !

Sure thing :)

> Labels might work for a few items but I'm looking at introducing a large number of preferences so this might not scale well with what I have in mind

Right. That makes sense, labels might not be the way to go. I wonder if we can find another way to reflect that at the VMI label in a way that scales. We can always introduce a new spec field, but I'm not sure how (and if) we can do it in a suitable way.

> ACK thanks, I'm still not convinced by this concern if I'm honest. Yes you would get different behaviour but these are after all different CRs. Wrapping VirtualMachine with some automation like this (and flavors, preferences etc) while leaving the runtime VirtualMachineInstance CR be more strict still seems like the right thing to do for me.

Well, that makes a whole lot of sense to me.
I don't think we must support this at the VMI level, but it obviously would be an advantage. IMHO we would see if it's possible to support VMIs in a way that makes sense, and if not, it's not a must-have.
 
> > > Additionally is it acceptable for us to heavily modify resources after submission like this?
> > 
> > 
> > Heavily modify which resource, the VMI? To which heavy modification are you referring - the injection of new disks to the VMI spec? How is it that different from this PR changes?
> 
> Yes the additional disks in this case but I'm also thinking in general of the application of preferences to the VMI post submission that I think you were also suggesting. The disks are added to the VMI prior to submission in this PR so the VMI resource itself doesn't change after submission.
> 
> > If a VMI is created by creating a VM -> there is no modification since we're the ones creating the VMI if the VMI is created directly -> we will only set default disks according to the defaults the VM creator set for us
> 
> In both cases the VMI resource is still going to be mutated after submission right? Even if we do that at the start of the VMI admittance process the definition supplied by the user or VM controller still isn't going to reflect what actually ends up running. Yes it will have references to things we apply but it isn't the same.
> 
> The reason I'm standing by the current proposal is because the users VM definition and the VMI definition created by the VM controller ultimately don't change after being admitted. We also leave all of the VMI validations in place to ensure that it can actually launch correctly.

Yes, that's a valid point.
On the one hand it's great to inject defaults and not enforce the user to define them all. On the other hand, many modifications can also be problematic if things get to crazy. For example, mutating webhooks are executed before validating webhooks. Massive changes at the mutation webhook level can be a shot to our own foot if in certain scenarios we mutate a VMI that would later wouldn't be validated, etc.

As I wrote above, this is not a must-have IMO, but we better discuss this to see if it can be done in a way that makes sense, because if we can support it that way it's clearly better.

@davidvossel I think we need your thoughts on this :)

Reply:
I think there are two dimensions here to untangle.

1. [this pr] automated disk defaulting when no disk is created that matches a volume.
2. [future pr] flavor hints for preferred disk bus that impacts disk defaulting.

For item 1, I think disk defaulting is a quality of the VMI. We already have a precedence this this kind of behavior in the VMI's mutation webhook. For example, a network and network interface is [implied if none exists](https://github.com/kubevirt/kubevirt/blob/4292ccf04262c5c612e6ea93d89f4e27ea961a6f/pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator.go#L90). Adding this disk defaulting to the mutation webhook aligns with existing behavior and would work naturally for VM and VMIs in a consistent way.

For item 2. Flavors are a quality of the VM. The expectation so far has been that the VM controller generates a VMI spec using the flavor and posts that result during vmi creation. So, once we have a preferred disk bus in the flavor, it will be the vm controller that generates the vmi's disk fields and injects the preferred disk bus if no bus is already present. The VMI's mutation webhook defaulter would then execute during that resulting VMI's creation to fill in missing gaps (like machinetype, missing default network, etc...)

Reply:
> I think there are two dimensions here to untangle.
> 
> 1. [this pr] automated disk defaulting when no disk is created that matches a volume.
> 2. [future pr] flavor hints for preferred disk bus that impacts disk defaulting.
> 
> For item 1, I think disk defaulting is a quality of the VMI. We already have a precedence this this kind of behavior in the VMI's mutation webhook. For example, a network and network interface is [implied if none exists](https://github.com/kubevirt/kubevirt/blob/4292ccf04262c5c612e6ea93d89f4e27ea961a6f/pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator.go#L90). Adding this disk defaulting to the mutation webhook aligns with existing behavior and would work naturally for VM and VMIs in a consistent way.

Thanks I completely missed AutoattachPodInterface, okay I'll rework this however as you go on to say below it's going to mean additional lookups of flavor preferences in the VMI mutation web hook.

> For item 2. Flavors are a quality of the VM. The expectation so far has been that the VM controller generates a VMI spec using the flavor and posts that result during vmi creation. So, once we have a preferred disk bus in the flavor, it will be the vm controller that generates the vmi's disk fields and injects the preferred disk bus if no bus is already present. The VMI's mutation webhook defaulter would then execute during that resulting VMI's creation to fill in missing gaps (like machinetype, missing default network, etc...)

ACK

Reply:
> > I think there are two dimensions here to untangle.
> > 
> > 1. [this pr] automated disk defaulting when no disk is created that matches a volume.
> > 2. [future pr] flavor hints for preferred disk bus that impacts disk defaulting.
> > 
> > For item 1, I think disk defaulting is a quality of the VMI. We already have a precedence this this kind of behavior in the VMI's mutation webhook. For example, a network and network interface is [implied if none exists](https://github.com/kubevirt/kubevirt/blob/4292ccf04262c5c612e6ea93d89f4e27ea961a6f/pkg/virt-api/webhooks/mutating-webhook/mutators/vmi-mutator.go#L90). Adding this disk defaulting to the mutation webhook aligns with existing behavior and would work naturally for VM and VMIs in a consistent way.
> 
> Thanks I completely missed AutoattachPodInterface, okay I'll rework this 

Hm, while I agree that we default there, I am still not sure if it is a similar type of defaulting. We default both sides, outside and inside and it has a clear scope (attach the default way to the pod network). I think a more matching equivalent would be a service account token which gets auto-attached than completing "half specs".

Also when reading this:

> however as you go on to say below it's going to mean additional lookups of flavor preferences in the VMI mutation web hook.

it looks to me like it could conceptually make sense to keep the flavour logic completely together in the VM controller and not distribute it.

It could also make sense to continue requiring conceptually complete  VMIs and only adjust it with self-sufficient additions like the pod network attachment. However this point about complete VMI specs is more a feeling than an argument backed by any real concerns at the moment.

Reply:
> Hm, while it is true that we default there, I am still not sure if it is a similar type of defaulting. We default both sides, outside and inside and it has a clear scope (attach the default way to the pod network).

yeah i get what you're saying. The difference between the network and volume defaulting is that we happen to have a default `network`, so it makes it easy to default the network + binding.  For volumes we don't know the volume part, but we can make some assumptions about the disk side if one isn't provided.

>it looks to me like it could conceptually make sense to keep the flavor logic completely together in the VM controller and not distribute it (less concerned about an extra API call, more about the concept).

yep, I think that's what we're going for.